scripts/src/java/org/oppia/android/scripts/wiki/WikiTableOfContentsCheck.kt

```diff
@@ -0,0 +1,100 @@
+package org.oppia.android.scripts.wiki
+
+import java.io.File
+
+/**
+ * Script for ensuring that the table of contents in each wiki page matches with its respective headers.
+ *
+ * Usage:
+ *   bazel run //scripts:wiki_table_of_contents_check -- <path_to_default_working_directory>
+ *
+ * Arguments:
+ * - path_to_default_working_directory: The default working directory on the runner for steps, and the default location of repository.
+ *
+ * Example:
+ *   bazel run //scripts:wiki_table_of_contents_check -- $(pwd)
+ */
+fun main(vararg args: String) {
+  // Path to the repo's wiki.
+  val wikiDirPath = "${args[0]}/wiki/"
+  val wikiDir = File(wikiDirPath)
+
+  // Check if the wiki directory exists.
+  if (wikiDir.exists() && wikiDir.isDirectory) {
+    processWikiDirectory(wikiDir)
+    println("WIKI TABLE OF CONTENTS CHECK PASSED")
+  } else {
+    println("No contents found in the Wiki directory.")
+  }
+}
+
+/**
+ * Checks every file in the wiki repo.
+ *
+ * @param wikiDir the default working directory
+ */
+fun processWikiDirectory(wikiDir: File) {
+  wikiDir.listFiles()?.forEach { file ->
+    checkTableOfContents(file)
+  }
+}
+
+/**
+ * Checks the contents of a single wiki file to ensure the accuracy of the Table of Contents.
+ *
+ * @param file the wiki file to process.
+ */
+fun checkTableOfContents(file: File) {
+  val fileContents = file.readLines()
+  val tocStartIdx = fileContents.indexOfFirst {
+    it.contains("## Table of Contents")
+  }
+  if (tocStartIdx == -1) {
+    println("No Table of Contents found for the file $file")
+    return
+  }
+
+  // Skipping the blank line after the ## Table of Contents
+  val eOfIdx = fileContents.subList(tocStartIdx + 2, fileContents.size).indexOfFirst {
+    it.isBlank()
+  }
+  if (eOfIdx == -1) error("Table of Contents didn't end with a blank line")
+
+  val tocSpecificLines = fileContents.subList(tocStartIdx, tocStartIdx + eOfIdx + 1)
+  println("Toc line: $tocSpecificLines")
+
+  for (line in tocSpecificLines) {
+    if (line.trimStart().startsWith("- [") && !line.contains("https://")) {
+      validateTableOfContents(file, line)
+    }
+  }
+}
+
+/**
+ * Validates the accuracy of a Table of Contents entry in a wiki file.
+ *
+ * @param file the wiki file being validated.
+ * @param line the line containing the Table of Contents entry.
+ */
+fun validateTableOfContents(file: File, line: String) {
+  val titleRegex = "\\[(.*?)\\]".toRegex()
+  val title = titleRegex.find(line)?.groupValues?.get(1)?.replace('-', ' ')
+    ?.replace(Regex("[?&./:’'*!,(){}\\[\\]+]"), "")
+    ?.trim()
+
+  val linkRegex = "\\(#(.*?)\\)".toRegex()
+  val link = linkRegex.find(line)?.groupValues?.get(1)?.removePrefix("#")?.replace('-', ' ')
+    ?.replace(Regex("[?&./:’'*!,(){}\\[\\]+]"), "")
+    ?.replace("confetti_ball", "")?.trim()
```

Comment:
Might we, instead, want to ban emojis? It would reduce the maintainance burden. WDYT?

Reply:
I have two thoughts on this but I'm unsure of the right approach:

1. We could restrict using emojis (But would this require adding a check to ensure no emojis are included in headers, or would it just be a manual review check?).
2. Alternatively, we could still accept emojis, but they should also be included in the TOC titles in the `:emoji:` form for comparison and validation.

Reply:
Wouldn't emojis fail by default since their presence would cause a mismatch?

Reply:
Ah, got it. I thought we should implement something to restrict adding emojis separately while pushing. In that case, the current check could work in both ways - by checking the TOC and also by allowing contributors to add emojis (if needed) but in a certain pattern.

- If an emoji is added as a visual emoji (standard image emoji) [🦦], the headers should ignore adding it. - [This is an option if contributors want to include emojis in their headers but don't want them in the TOC representation]
- If an emoji is added as GitHub emoji code `(:emoji:)` [:otter:], the headers should include it. - [This is an option if contributors want to represent them in both headers and the TOC representation]

Few examples that would work are:

### Table of Contents
- [Instructions1](#instructions1-)
- [Instructions2 :otter:](#instructions2-otter)
- [Instructions3](#-instructions3)
- [:otter: Instructions4](#otter-instructions4)

```
- [Instructions1](#instructions1-)
- [Instructions2 :otter:](#instructions2-otter)
- [Instructions3](#-instructions3)
- [:otter: Instructions4](#otter-instructions4)
```

### Headers
- Instructions1 🦦
- Instructions2 :otter:
- 🦦 Instructions3
- :otter: Instructions4

```
## Instructions1 🦦
## Instructions2 :otter:
## 🦦 Instructions3
## :otter: Instructions4
```

So, anything in the wrong format would be checked by the script, and contributors can decide how they want to represent it, I guess.

Reply:
Yes, exactly. Also resolving since we agreed in the meeting discussion.